types/types.cc

```diff
@@ -1610,6 +1632,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
     }
 }
 
+sstring vector_type_impl::make_name(data_type type, size_t dimension) {
+    // To keep format compatibility with Origin we never wrap
+    // vector name into
+    // "org.apache.cassandra.db.marshal.FrozenType(...)".
+    return seastar::format("org.apache.cassandra.db.marshal.VectorType({}, {})", type->name(), dimension);
+}
+
+vector_type_impl::vector_type_impl(data_type elements, size_t dimension)
+        : concrete_type(kind::vector, make_name(elements, dimension),
+        elements->value_length_if_fixed() ? std::optional(elements->value_length_if_fixed().value()*dimension):std::nullopt),
+        _elements_type(elements), _dimension(dimension) {
+    _contains_set_or_map = _elements_type->contains_set_or_map();
+}
+
+shared_ptr<const vector_type_impl>
+vector_type_impl::get_instance(data_type elements, size_t dimension) {
+    return intern::get_instance(elements, dimension);
+}
+
+static void serialize_vector(const vector_type_impl& type, const vector_type_impl::native_type* val, bytes::iterator& out) {
+    auto elements_type = type.get_elements_type();
+    if (type.value_length_if_fixed()) {
+        for (const auto& value : *val) {
+            value.serialize(out);
+        }
+    } else {
+        for (const auto& value : *val) {
+            size_t val_len = value.serialized_size();
+
+            out += unsigned_vint::serialize(val_len, out);
+            
+
+            value.serialize(out);
+        }
+    }
+}
+
+std::strong_ordering
+vector_type_impl::compare_vectors(data_type elements, size_t dimension, managed_bytes_view o1, managed_bytes_view o2) {
+    if (o1.empty()) {
+        return o2.empty() ? std::strong_ordering::equal : std::strong_ordering::less;
+    } else if (o2.empty()) {
+        return std::strong_ordering::greater;
+    }
+
+    for (size_t i = 0; i < dimension; i++) {
+        auto v1 = read_vector_element(o1, elements->value_length_if_fixed());
+        auto v2 = read_vector_element(o2, elements->value_length_if_fixed());
+        auto cmp = elements->compare(v1, v2);
+        if (cmp != 0) {
+            return cmp;
+        }
+    }
+
```

Comment:
Please verify that this ordering is compatible with Cassandra.

Reply:
This is why I encourage people to write all tests that can be exercised through CQL in test/cqlpy and not in the test/boost cql_env framework. The latter always leaving you wondering whether the behavior you just wrote a test for actually matches Cassandra's behavior. The former allows you to really be sure that Scylla's behavior is identical to Cassandra and not incompatible with Cassandra for no good read.

I saw you translated Cassandra's unit tests and put it in test/cqlpy/cassandra_tests, which is great, but you can also add new cqlpy tests, just don't put them in cassandra_tests (a new file test/cqlpy/test_type_vector.py would be a good place for new tests for the vector type).